bsmetadata/preprocessing_tools/html_parser/filters_and_cleaners.py

```diff
@@ -0,0 +1,459 @@
+from bsmetadata.preprocessing_tools.html_parser.variables import (
+    BLOCK_CONTENT_SEPARATOR,
+    BLOCK_ELEMENTS,
+    FAKE_TAG_BASIC,
+    FAKE_TAG_BLOCK,
+    FAKE_TAG_INLINE,
+    INLINE_ELEMENTS_SPACING,
+    PLAIN_TEXT_SEPARATOR,
+    PRE_TAG,
+)
+
+
+class AttributeCleaner:
+    def __init__(self, attrs_to_keep: Optional[List[str]]):
+        self.attrs_to_keep = attrs_to_keep
+
+    def _test(self, attr):
+        return self.attrs_to_keep is None or attr in self.attrs_to_keep
+
+    def __call__(self, attrs: List[Tuple[str]]):
+        if isinstance(attrs, list):
+            attrbs = [attr for attr, _ in attrs if self._test(attr)]
+            values = [value for attr, value in attrs if self._test(attr)]
+            return {
+                "attrs": attrbs,
+                "values": values,
+            }
+        else:
+            attrs = dict(attrs)
+
+            attrbs = [attr for attr, value in attrs.items() if self._test(attr)]
+            values = [value for attr, value in attrs.items() if self._test(attr)]
+            return {
+                "attrs": attrbs,
+                "values": values,
+            }
+
+
+class TagFilter:
+    def __init__(
+        self,
+        tags_to_remove_alone: Optional[List[TagToRemove]],
+        tags_to_remove_with_content: Optional[List[TagToRemoveWithContent]],
+        txt_max_chr_len_alone: Optional[float] = -float("inf"),
+        txt_min_chr_len_alone: Optional[float] = -float("inf"),
+        tags_exceptions_alone: Optional[List[str]] = None,
+        txt_max_chr_len_with_content: Optional[float] = -float("inf"),
+        txt_min_chr_len_with_content: Optional[float] = -float("inf"),
+        tags_exceptions_with_content: Optional[List[str]] = None,
+    ):
+        self.txt_max_chr_len_alone = txt_max_chr_len_alone
+        self.txt_min_chr_len_alone = txt_min_chr_len_alone
+        self.tags_exceptions_alone = tags_exceptions_alone if tags_exceptions_alone is not None else []
+        self.txt_max_chr_len_with_content = txt_max_chr_len_with_content
+        self.txt_min_chr_len_with_content = txt_min_chr_len_with_content
+        self.tags_exceptions_with_content = (
+            tags_exceptions_with_content if tags_exceptions_with_content is not None else []
+        )
+
+        self.tags_to_remove_alone = (
+            {tag_to_remove.tag: tag_to_remove for tag_to_remove in tags_to_remove_alone}
+            if isinstance(tags_to_remove_alone, list)
+            else {}
+        )
+        self.tags_to_remove_with_content = (
+            {tag_to_remove.tag: tag_to_remove for tag_to_remove in tags_to_remove_with_content}
+            if isinstance(tags_to_remove_with_content, list)
+            else {}
+        )
+
+        for tag_to_remove_characteristics in self.tags_to_remove_with_content.values():
+            if tag_to_remove_characteristics.method not in ["top-down", "bottom-up"]:
```

Comment:
Could you briefly explain the difference between top-down and bottom-up removal? I'm not sure I fully understand the difference from looking at the code.

Reply:
![image](https://user-images.githubusercontent.com/55560583/143036886-a15ab9da-293b-4efb-8f33-7aca44f23763.png)

Of course! Tags can be removed from the HTML DOM based on the length of the text contained in the tag.  

The "top-down" method will consist in going through the DOM from the root to the leaves and if during this traversal a node fulfills the conditions of elimination it will be deleted.

On the contrary, the "bottom-up" method will consist in going through the DOM from the leaves to the root and if during this traversal a node meets the conditions of elimination it will be deleted.

I implemented both but I think "bottom-up"  is a bit weird (by default "top-down"  is selected). I hesitated to delete  "bottom-up" but I thought that it was better not to lose the code if ever during the discussion this method appears to be useful.